util/chunk/chunk.go

```diff
@@ -30,24 +32,60 @@ type Chunk struct {
 	// numVirtualRows indicates the number of virtual rows, which have zero column.
 	// It is used only when this Chunk doesn't hold any data, i.e. "len(columns)==0".
 	numVirtualRows int
+	// capacity indicates the max number of rows that the chunk wants to hold.
+	// The accurate row count after executor.Next is guaranteed by executor implement.
+	// It is used only during executor's execution, and grow by `Renew` or `Reset`.
+	capacity int
 }
 
 // Capacity constants.
 const (
 	InitialCapacity = 32
+	varElemLen      = -1
 )
 
 // NewChunkWithCapacity creates a new chunk with field types and capacity.
 func NewChunkWithCapacity(fields []*types.FieldType, cap int) *Chunk {
+	return New(fields, cap, cap) //FIX ME in continue PR, this keep every thing same.
+}
+
+// New creates a new chunk.
+//  cap: the max number of rows that the chunk wants to hold.
+//  capLimit: the max limit for max number of rows.
+func New(fields []*types.FieldType, cap, capLimit int) *Chunk {
 	chk := new(Chunk)
 	chk.columns = make([]*column, 0, len(fields))
-	chk.numVirtualRows = 0
 	for _, f := range fields {
-		chk.addColumnByFieldType(f, cap)
+		chk.columns = append(chk.columns, newColumn(getFixedLen(f), cap, nil))
 	}
+	chk.numVirtualRows = 0
+	chk.capacity = mathutil.Min(chk.capacity, capLimit)
 	return chk
 }
 
+// Renew recreates a new chunk base on old chunk.
+//   chk: old chunk(often used in previous call).
+//   capLimit: the max limit for max number of rows.
+//
+//  this method will be used in call Executor#Next many times with a new chunk situation.
+//  so it will `checkCapacity` to get a new cap then create a new chunk, whatever happens.
+func Renew(chk *Chunk, capLimit int) *Chunk {
+	_, newCap := checkCapacity(chk, capLimit, true)
+	newChk := new(Chunk)
+	newChk.columns = resizeColumns(chk.columns, newCap)
+	newChk.numVirtualRows = 0
+	newChk.capacity = newCap
+	return newChk
+}
+
+func resizeColumns(oldCol []*column, cap int) []*column {
```

Comment:
Why not name it `renewColumns`?
`resize` implies modification on the old columns.